src/main/java/net/pms/dlna/DLNAThumbnailInputStream.java

```diff
@@ -0,0 +1,348 @@
+	 * @return The populated {@link DLNAThumbnailInputStream} or {@code null}
+	 *         if the source image could not be parsed.
+	 * @throws IOException if the operation fails.
+	 */
+	public static DLNAThumbnailInputStream toThumbnailInputStream(
+		InputStream inputStream,
+		int width,
+		int height,
+		ScaleType scaleType,
+		ImageFormat outputFormat,
+		boolean padToSize
+	) throws IOException {
+		DLNAThumbnail thumbnail = DLNAThumbnail.toThumbnail(inputStream, width, height, scaleType, outputFormat, padToSize);
+		return thumbnail != null ? new DLNAThumbnailInputStream(thumbnail) : null;
+	}
+
+    /**
+     * Creates a {@link DLNAThumbnailInputStream} where it uses
+     * {@code thumbnail}'s buffer as its buffer array. The buffer array is
+     * not copied.
+     *
+     * @param thumbnail the input thumbnail.
+	 * @return The populated {@link DLNAThumbnailInputStream} or {@code null}
+	 *         if the source thumbnail is {@code null}.
+     */
+
+	public static DLNAThumbnailInputStream toThumbnailInputStream(DLNAThumbnail thumbnail) {
+		return thumbnail != null ? new DLNAThumbnailInputStream(thumbnail) : null;
+	}
+
+	/**
+	 * Don't call this from outside this class or subclass, use
+	 * {@link DLNAThumbnailInputStream#toThumbnailInputStream(DLNAThumbnail)}
+	 * which handles {@code null} input.
+	 *
+	 * @param thumbnail the input thumbnail
+	 *
+	 * @throws NullPointerException if {@code thumbnail} is {@code null}.
+	 */
+    protected DLNAThumbnailInputStream(DLNAThumbnail thumbnail) {
+    	super(thumbnail.getBytes(false));
+    	this.imageInfo = thumbnail.getImageInfo();
+    	this.profile = thumbnail.getDLNAImageProfile();
+    }
+
+	/**
+	 * Converts and scales a thumbnail according to the given
+	 * {@link DLNAImageProfile}. Preserves aspect ratio. Format support is
+	 * limited to that of {@link ImageIO}.
+	 *
+	 * @param outputProfile the DLNA media profile to adhere to for the output.
+	 * @param updateMetadata whether or not new metadata should be updated
+	 *                       after image transformation. This should only be
+	 *                       disabled if the output image won't be kept/reused.
+	 * @param padToSize Whether padding should be used if source aspect doesn't
+	 *                  match target aspect.
+	 * @return The scaled and/or converted thumbnail, {@code null} if the
+	 *         source is {@code null}.
+	 * @exception IOException if the operation fails.
+	 */
+	public DLNAThumbnailInputStream transcode(
+		DLNAImageProfile outputProfile,
+		boolean updateMetadata,
+		boolean padToSize
+	) throws IOException {
+		DLNAThumbnail thumbnail;
+		thumbnail = (DLNAThumbnail) ImagesUtil.transcodeImage(
+			this.getBytes(false),
+			outputProfile,
+			updateMetadata,
+			true,
+			padToSize);
+		return thumbnail != null ? new DLNAThumbnailInputStream(thumbnail) : null;
+	}
+
+	/**
+	 * @return The bytes of this thumbnail.
+	 */
+	@SuppressFBWarnings("EI_EXPOSE_REP")
```

Comment:
In that exact example, it's because Findbugs will complain about returning ```buf``` is ```copy``` is false below:
```java
	public byte[] getBytes(boolean copy) {
		if (copy) {
			byte[] result = new byte[buf.length];
			System.arraycopy(buf, 0, result, 0, buf.length);
			return result;
		} else {
			return buf;
		}
	}
``` 
```buf``` is a protected byte array held (referenced) by an instance of ```DLNAThumbnailInputStream```. By returning this reference to a caller, you actually return the reference to the byte array in memory. If that caller then modifies any values in that byte array, what is being helt by the ```DLNAThumbnailInputStream``` is also modified. That's what they mean by "exposing internal representation". 

Even though in theory I'd like to avoid doing that, the only way to do that in reality is to copy everything being received or sent by the instance. While that makes it "tamper safe", it will slow everything down and require more memory. What if being held in the byte array here is actually a whole image in JPEG or PNG format, so copying it could be expensive - and so is holding the same data multiple times in memory. 

The thing is that the code isn't actually modifying the array anywhere. It's a bit complicated, but since an array is actually just a mapping to a part of the memory that is allocated (marked as non-free), it has a fixed size. You can't change the size of an array once it's created, if you need to do that you'll have to create a new one with the new size and then copy parts of or all of the "old" array to the new array (and add any additional information you might have). The old array is then discarded (or, in Java it isn't actually, it's just made available for the garbage collector to free at some unknown point in the future). Because practically any modification to an image will let to some change in size, you never actually reuse the array. You always create a new one when you make changes. As such, the likelihood of the array actually being modified in close to none.

That said, I wish there was such a thing as an immutable array in Java (an array you can't write to after it's creation). I'd use that exclusively, but sadly there isn't. So, I'm doing as I am here, taking a little bit of risk to gain performance/save memory.

It's nice that Findbugs warns though, it's bad if you do something like this unintentionally. But, when it *is* intentional I simply suppress the warning so it doesn't make it harder to spot the warnings I want to see.